### ISSUE TITLE
[feature request] A rank-revealing SVD for better stability in backward.

### ISSUE THREAD

**Issue by nikitaved** [pytorch/pytorch#69532]
## 🚀 Feature

It would be great to have a rank-revealing SVD decomposition which, for a given input of size `m x n` returns factors of rank `r` with `r <= min(m, n)`.

## Motivation

The main motivation is backward stability. The current implementation of SVD makes backward unstable when the input has repeated and/or zero singular values. See, for example: https://github.com/pytorch/pytorch/issues/69531. This kills the possibility to use SVD.backward on low-rank inputs which sure have (repeated) zero singular values.

`torch.linalg.svd` could support additional parameters (like `tol`, and the name for the rank-revealing code path) to enable this feature.
No need to modify the `svd_backward`, as with reduced outputs the chance of backpropagating through zero singular values is significantly reduced.

**Comment by redwrasse** [comment]
Hey @nikitaved , any appetite for this several years on? Do you think it's still a valuable addition to the Pytorch codebase at its current state?

**Comment by ezyang** [comment]
I don't fundamentally have any opposition to shepherding this but we are a bit thin on linear algebra experts right now, so the burden would be on the PR writer to convince non expert reviewers that the PR was correct and inevitable

**Comment by redwrasse** [comment]
@nikitaved I don't know enough up-front to comment on the design choices. Happy to follow your lead and the non-merged PR sounds like a good place to start. It sounds like avoiding syncs and implementing out of core are what you'd like to target?

**Comment by redwrasse** [comment]
@nikitaved https://github.com/pytorch/pytorch/pull/71126 is the non-merged branch, correct?

**Comment by nikitaved** [comment]
@redwrasse , yes, this is the one! Sorry for the delays in my replies. Would you like to work on it?
